HADOOP-19330. S3A: Add LeakReporter; use in S3AInputStream

steveloughran opened the pull request:
S3AInputStream.finalizer() to
- abort any active HTTP connection.
- warn if there was one
- warning to include: path, and stack+thread ID of creation.

The warning message is logged at WARN to log

   org.apache.hadoop.fs.s3a.connection.leaks

set to a lower log level to hide these messages

This is a best-effort recovery from stream leaks.
Cleanup will *only* take place during GC; it is easy to run out of connections without running out of memory.


### How was this patch tested?

New ITest.

### For code changes:

- [X] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?
- [X] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

steveloughran committed:
HADOOP-19330. S3A: S3AInputStream to release http connection on GC

S3AInputStream.finalizer to
- abort any active HTTP connection.
- warn if there was one
- warning to include: path, and stack+thread ID of creation.

The warning message is logged at WARN to log

   org.apache.hadoop.fs.s3a.connection.leaks

set to a lower log level to hide these messages

This is a best-effort recovery from stream leaks.
Cleanup will *only* take place during GC; it is easy
to run out of connections without running out of memory.

Change-Id: I162c347ce292a5e7e053a03342bf69b7bca17b11

steveloughran commented:
what the log looks like

S3AInputStream
```
2024-11-07 17:50:36,378 [Finalizer] DEBUG s3a.S3AInputStream (S3AInputStream.java:closeStream(759)) - Closing stream finalize(): abort
2024-11-07 17:50:36,378 [Finalizer] DEBUG impl.SDKStreamDrainer (SDKStreamDrainer.java:drainOrAbortHttpStream(176)) - drain or abort reason finalize() remaining=1023 abort=true
2024-11-07 17:50:36,378 [Finalizer] DEBUG impl.SDKStreamDrainer (SDKStreamDrainer.java:drainOrAbortHttpStream(235)) - Aborting stream s3a://stevel-london/test/testFinalizer
2024-11-07 17:50:36,380 [Finalizer] DEBUG impl.SDKStreamDrainer (SDKStreamDrainer.java:drainOrAbortHttpStream(245)) - Stream s3a://stevel-london/test/testFinalizer aborted: finalize(); remaining=1023
```

new warning log
```

2024-11-07 17:50:36,380 [Finalizer] WARN  connection.leaks (S3AInputStream.java:finalize(292)) - HTTP connection not closed while reading s3a://stevel-london/test/testFinalizer in thread JUnit-testFinalizer
java.io.IOException: HTTP connection not closed while reading s3a://stevel-london/test/testFinalizer in thread JUnit-testFinalizer
	at org.apache.hadoop.fs.s3a.S3AInputStream.<init>(S3AInputStream.java:263)
	at org.apache.hadoop.fs.s3a.S3AFileSystem.executeOpen(S3AFileSystem.java:1890)
	at org.apache.hadoop.fs.s3a.S3AFileSystem.open(S3AFileSystem.java:1840)
	at org.apache.hadoop.fs.FileSystem.open(FileSystem.java:997)
	at org.apache.hadoop.fs.s3a.ITestS3AInputStream.testFinalizer(ITestS3AInputStream.java:60)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:750)
```

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 23s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  32m  9s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 26s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  compile  |   0m 22s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  checkstyle  |   0m 23s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 30s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   0m 27s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   0m 22s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  spotbugs  |   0m 46s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  20m 12s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 17s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 22s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javac  |   0m 22s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 17s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  javac  |   0m 17s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   0m 14s | [/results-checkstyle-hadoop-tools_hadoop-aws.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/1/artifact/out/results-checkstyle-hadoop-tools_hadoop-aws.txt) |  hadoop-tools/hadoop-aws: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
| +1 :green_heart: |  mvnsite  |   0m 19s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 19s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   0m 18s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| -1 :x: |  spotbugs  |   0m 49s | [/new-spotbugs-hadoop-tools_hadoop-aws.html](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/1/artifact/out/new-spotbugs-hadoop-tools_hadoop-aws.html) |  hadoop-tools/hadoop-aws generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
| +1 :green_heart: |  shadedclient  |  20m  5s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 19s |  |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 26s |  |  The patch does not generate ASF License warnings.  |
|  |   |  82m 32s |  |  |


| Reason | Tests |
|-------:|:------|
| SpotBugs | module:hadoop-tools/hadoop-aws |
|  |  org.apache.hadoop.fs.s3a.S3AInputStream.finalize() is public; should be protected  At S3AInputStream.java:protected  At S3AInputStream.java:[lines 288-295] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/7151 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 3e151f25475d 5.15.0-124-generic #134-Ubuntu SMP Fri Sep 27 20:20:17 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 0841b7eae3c8fe6122d0b00f6603d3d209ad3539 |
| Default Java | Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/1/testReport/ |
| Max. process+thread count | 726 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

steveloughran committed:
HADOOP-19330 documentation + yetus

Change-Id: Id7808e1e178c50cfa6e201d6b28a8827573696a7

steveloughran commented:
This is ready for review but a tricky one to test.

Anyone who understands GC and finalizers is very welcome to look at this and provide feedback.

steveloughran commented:
This is ready for review but a tricky one to test. The test *does* work but it took me awhile to realise that we are finalizers our run in a separate thread and you need to put a sleep in so the captured log does not to come back empty.

Anyone who understands GC and finalizers is very welcome to look at this and provide feedback. 

@mukund-thakur @saikatroy038 @HarshitGupta11 @ahmarsuhail @shameersss1

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 31s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +0 :ok: |  markdownlint  |   0m  1s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  44m 29s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 46s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  compile  |   0m 36s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  checkstyle  |   0m 33s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 43s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   0m 43s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   0m 34s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  spotbugs  |   1m 10s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  35m 19s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 29s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 36s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javac  |   0m 36s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 27s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  javac  |   0m 27s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 20s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 32s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 28s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   0m 27s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  spotbugs  |   1m  9s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  35m  9s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   3m  0s |  |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 38s |  |  The patch does not generate ASF License warnings.  |
|  |   | 129m 45s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/7151 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets markdownlint |
| uname | Linux 7caad80554cd 5.15.0-124-generic #134-Ubuntu SMP Fri Sep 27 20:20:17 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 16ad1352617a8459a83b6815ff5db45800cd9039 |
| Default Java | Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/2/testReport/ |
| Max. process+thread count | 754 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/2/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

steveloughran commented:
I did just write a marginal optimisation to only build the stack trace if the log was <= warn but haven't committed it. It just wasn't testable and I did not want to create the risk of an NPE for marginal speedups. These objects are talking over HTTPS to object storage; the overhead of stack treewalk is insignifcant.

steveloughran committed:
HADOOP-19330. Stream finalize.

Pull out leak reporting into hadoop-common class
org.apache.hadoop.fs.impl.LeakReporter implements Closeable

Which takes
-message to print
-predicate for being open
-RunnableRaisingIOE (new class) for the close action.

+ to string, re-entrancy guard, catch and swallow of failures...

Has a hadoop-common unit test.

Change-Id: Iaee47fc26e674d29163b6ffc7b598187e44a7675

ahmarsuhail reviewed (approved):
I don't have a good understanding of finalizers or GC, but the code change itself looks good to me :)

steveloughran commented:
@ahmarsuhail thanks. my last commit actually pulls out everything into a hadoop-common `LeakReporter` class which can be tested on its own; it can then be used by anything we need close()ing which applications may forget to to.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 32s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  markdownlint  |   0m  0s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 2 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  15m 16s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  31m 54s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  17m 46s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  compile  |  16m 11s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  checkstyle  |   4m 20s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 43s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   2m 11s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   1m 45s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  spotbugs  |   4m  1s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  35m 47s |  |  branch has no errors when building and testing our client artifacts.  |
| -0 :warning: |  patch  |  36m 14s |  |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 34s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 32s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 47s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javac  |  16m 47s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 16s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  javac  |  16m 16s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   4m 21s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/3/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
| +1 :green_heart: |  mvnsite  |   2m 38s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   2m  6s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   1m 46s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  spotbugs  |   4m 17s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  35m 33s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  |  20m 44s | [/patch-unit-hadoop-common-project_hadoop-common.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/3/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt) |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   3m 14s |  |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  5s |  |  The patch does not generate ASF License warnings.  |
|  |   | 247m 23s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.fs.impl.TestLeakReporter |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/7151 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets markdownlint |
| uname | Linux 4da2f74f65c3 5.15.0-124-generic #134-Ubuntu SMP Fri Sep 27 20:20:17 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 74496c275314cfaf1c9b07abb075ca1dc8dabadb |
| Default Java | Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/3/testReport/ |
| Max. process+thread count | 1251 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/3/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

steveloughran committed:
HADOOP-19330. Stream Leaks

Add a Stream statistic for this with

* Capability of FS (if !prefetching) and stream itself (tested)
* statistic collected in stream and merged with FS (tested)
* test tuning (with failing test fixed)
* doc review

Change-Id: I747a9477e0571d729c81e6f17dac127ffb6ade60

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 36s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  markdownlint  |   0m  0s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 2 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  14m 44s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  38m  5s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  20m 26s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  compile  |  19m 11s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  checkstyle  |   5m  7s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   3m  8s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   2m  3s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   1m 38s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  spotbugs  |   4m  5s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  39m 25s |  |  branch has no errors when building and testing our client artifacts.  |
| -0 :warning: |  patch  |  39m 54s |  |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 31s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 32s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  17m 46s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javac  |  17m 46s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 52s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  javac  |  16m 52s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   4m 26s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/4/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 1 new + 4 unchanged - 0 fixed = 5 total (was 4)  |
| +1 :green_heart: |  mvnsite  |   2m 37s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   2m  6s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   1m 43s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  spotbugs  |   4m 28s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  39m 24s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  21m  3s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   3m  9s |  |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  7s |  |  The patch does not generate ASF License warnings.  |
|  |   | 269m 15s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/7151 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets markdownlint |
| uname | Linux 8866b2cd903b 5.15.0-124-generic #134-Ubuntu SMP Fri Sep 27 20:20:17 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 9daa3ee04fee1e5e81dd92eb3f60f770c0590349 |
| Default Java | Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/4/testReport/ |
| Max. process+thread count | 3151 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/4/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

mukund-thakur commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java:
```diff
@@ -5593,6 +5594,10 @@ public boolean hasPathCapability(final Path path, final String capability)
       return s3AccessGrantsEnabled;
 
+      // stream leak detection.
+    case StreamStatisticNames.STREAM_LEAKS:
```
disabled because we will have multiple prefetch streams reading data ?

steveloughran replied:
prefetch doesn't detect leaks, so this capability is off then

mukund-thakur commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AInputStreamLeakage.java:
```diff
@@ -0,0 +1,161 @@
+      // remove strong reference to the stream
+      in = null;
+      // force the gc.
+      System.gc();
```
this doesn't necessarily always trigger GC but is best effort. 
I remember working on similar tests but now you have found runFinalization. Never used this. 
but isn't gc supposed to run finalize anyway?  
Reading the java doc I think gc will do that but this will ensure it is done.

steveloughran replied:
It's hard to test as you have to get rid of all references; the weakref reference lets us assert that this happened. Maybe race conditions in future -we will have to see.

Using the callback to update a counter that was verified that the finalizer ran

mukund-thakur commented on hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/impl/TestLeakReporter.java:
```diff
@@ -0,0 +1,181 @@
+        .describedAs("close count")
+        .isEqualTo(ex);
+  }
+}
```
nit: checkstyle blank line

mukund-thakur commented on hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/connecting.md:
```diff
@@ -485,6 +485,90 @@ If `storediag` doesn't connect to your S3 store, *nothing else will*.
+it will log at `WARN` level that it has not been closed,
+listing the file URL, and the thread name + ID of the the thread
+which creating the file.
+The the stack trace of the `open()` call will be logged at `INFO`
```
wondering where and how the stack trace of open is getting printed?

mukund-thakur replied:
NM, got it. since it is getting initialized in S3AInputStream.

steveloughran replied:
I considered only collecting the stack if the specific log was at error, but decided it was just adding risk of failures and hard to test. The extra overhead is marginal given we are about to do network reads.

Also gives us a full stack during any debugging.

Now, one future enhancement would be to include that in the .toString() of the reporter, and then include that in the .toString() of the stream. Just a thought. Not doing it here

mukund-thakur reviewed (commented):
LGTM +1 pending yetus

steveloughran committed:
HADOOP-19330. Stream Leaks

Style

Change-Id: I5ecc157500bfb84d94a40d712df45994b7dddb3b

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 38s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +0 :ok: |  markdownlint  |   0m  1s |  |  markdownlint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 2 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  14m 38s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  32m  2s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  17m 30s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  compile  |  16m 11s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  checkstyle  |   4m 25s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 39s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   2m 11s |  |  trunk passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   1m 45s |  |  trunk passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  spotbugs  |   3m 57s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  36m 54s |  |  branch has no errors when building and testing our client artifacts.  |
| -0 :warning: |  patch  |  37m 21s |  |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 32s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 29s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  17m  9s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javac  |  17m  9s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 20s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  javac  |  16m 20s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   4m 22s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   2m 38s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   2m  8s |  |  the patch passed with JDK Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04  |
| +1 :green_heart: |  javadoc  |   1m 45s |  |  the patch passed with JDK Private Build-1.8.0_422-8u422-b05-1~20.04-b05  |
| +1 :green_heart: |  spotbugs  |   4m 21s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  36m  3s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  20m 47s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   3m 11s |  |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  5s |  |  The patch does not generate ASF License warnings.  |
|  |   | 248m 54s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/7151 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets markdownlint |
| uname | Linux 9c4c4f44e615 5.15.0-124-generic #134-Ubuntu SMP Fri Sep 27 20:20:17 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / d13f1f20a5a4baff9b22d44f3eef2274f1ed730a |
| Default Java | Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.24+8-post-Ubuntu-1ubuntu320.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_422-8u422-b05-1~20.04-b05 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/5/testReport/ |
| Max. process+thread count | 3151 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7151/5/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.
